symfony/symfony
# [Messenger] use Envelope internally, return void, add EnvelopeHandlerInterface and other cleanups

| Q             | A
| ------------- | ---
| Branch?       | 4.2
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | not yet
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Sorry in advance for this late PR, I've just reviewed the component for the first time and I think we should do some changes in its internal design, so here we are, last chance before tagging it as stable.

Each item in this list can potentially trigger a discussion. Here are the topics of this PR:
- make everything *internally* deal with `Envelope` objects (middlewares, transports, etc.)
- ~replace `EnvelopeAwareInterface` (for middlewares) by `EnvelopeHandlerInterface` (for userland handlers)~ [edit: bad idea, see discussion]
- remove the return value of `dispatch()`+`handle()`: we never use it and it contradicts the very idea of async (mutable messages can be used instead, but that's not our business)
- added `string $name = null` as 2nd argument to dispatch(), used as dispatching key when provided for more flexibility (it's not always possible/desirable to create a new class to define a dispatching key)
- ~locating senders/handlers using the parent classes+interfaces of the message is removed.~ To me, locating based on the type hierarchy is a code smell that we hijack the type hierarchy to make it fit what should be done via configuration. [edit: now kept to reduce the impact of this PR - could be deprecated later if this smell is proven by experience in the future]
- remove `AllowNoHandlerMiddleware` and replace it with a constructor argument on `HandleMessageMiddleware`: using exceptions for signaling is going to slow down the component for no real reasons to me. This solves it. 
- rename some methods / change some interfaces for consistency/ease of use

 I expect this to raise some discussions. To resolve them one by one, I propose to split this PR into separate ones while we move our common understanding forward.

Meanwhile, I didn't update tests nor DI integration. I'll do as this moves forward.
